### ISSUE TITLE
Two simple plugins - should they be added to the common lib?

### ISSUE THREAD

**Issue by jason-fox** [telefonicaid/iotagent-node-lib#879]
I'm working on a new IoT Agent where the input format is fixed but I want to follow standard data models within my context data. I have created a couple of plugins to solve my current problem and I'm wondering if the solution code is common enough to warrant a PR on the library (and then adding them to the `addUpdateMiddleware()` list in each plugin).

For example. Imagine I was using **Ultralight** and I wanted to push an `address` **Object** when I receive a measure:

e.g. the following payload

```text
B|Manor Farm|C|19  Stamford Road|E|OX13 2ZY|F|APPLETON|G|Oxfordshire
```

What I really want to end up with is just

```json
{
        "id": "Farm:FRM3",
        "type": "Farm",
        "TimeInstant": "2020-07-03T15:33:29.00Z",
        "address": {
            "street": "19  Stamford Road",
            "city": "APPLETON",
            "state": "Oxfordshire",
            "postalCode" : "OX13 2ZY"
        },
        "name": "Manor Farm"
    }
```

Now, I can directly map `B => name` which is a **String**, (and `B` is not included as a separate attribute)  but I can't currently create a full JSON object (or am I missing something with the existing expressions plugin). Anyway the best I can do at the moment for the `address` attribute is to create the value using simple strings expressions and have unnecessary attributes on my context data:

```json
{
        "id": "Farm:FRM3",
        "type": "Farm",
        "C": "19  Stamford Road",
        "E": "OX13 2ZY",
        "F": "APPLETON",
        "G": "Oxfordshire",
        "TimeInstant": "2020-07-03T15:33:29.00Z",
        "address": "19  Stamford Road, APPLETON, OX13 2ZY, Oxfordshire",
        "name": "Farm2"
}
```

So I've created two 10 line plugins to do the following:

1. transform an expression string into a JSON object

```javascript
function update(entity, typeInformation, callback) {
	Object.keys(entity).forEach((key) => {
		const value = entity[key].value;
		if (typeof value === 'string' || value instanceof String){
			if (value.startsWith('{') && value.endsWith('}')){
				entity[key].value = JSON.parse(value);
			}
		}
	});
    return callback(null, entity, typeInformation);
}
```

2. delete a given set of attributes I don't want in my NGSI entity.

```javascript
function update(entity, typeInformation, callback) {
    const internals = typeInformation.internalAttributes;
    if(internals && Array.isArray(internals) && internals.length > 0){
    	const deleteAttrs = internals[0].delete;
    	if(deleteAttrs){
    		deleteAttrs.forEach((attr)=>{
    			delete entity[attr];
    		});
    	}
    }
    return callback(null, entity, typeInformation);
}
```

This means I can provision my service group like this:

```console
curl -L -X POST 'http://localhost:4041/iot/services' \
-H 'Content-Type: text/plain' \
-H 'Cookie: _csrf=MAPTGFPcoPnewsGCWklHi4Mq' \
--data-raw '{
 "services": [
   {
... etc
     "attributes": [
       { "object_id": "B", "name":"name", "type":"Text"},
       {
               "name":"address",
               "type":"PostalAddress",
               "expression": "{ \"x\": \"${@C}\",\"y\": \"${@F}\"}"
        }
      ],
      "internal_attributes":{
      	"delete" : ["A", "B", "C", "D", "E", "F"]
      }
   }
 ]
}'
```

And send data in the format I need. Would these plugins be useful for other developers?

**Comment by AlvaroVega** [comment]
@jason-fox  What `expressionLanguage are you using for that example?

**Comment by fgalan** [comment]
> ...  but I can't currently create a full JSON object (or am I missing something with the existing expressions plugin)

I understood that using the recently contributed JEXL expression plugin this can be solved, but maybe I'm wrong. @chicco785 what do you think? How it could be the solution to @jason-fox 's use case using JLEX?

**Comment by fgalan** [comment]
With regards to attribute deletion, maybe the selective ignore of measures (also recently contributed) could be used (explicitAttrs: true at provision time)?

**Comment by chicco785** [comment]
i don't think @jason-fox is using jexl

https://github.com/telefonicaid/iotagent-node-lib/blob/d47127ae817c940080c4517bb6aedaef54576f48/doc/expressionLanguage.md#examples-of-expressions-1

**Comment by mrutid** [comment]
I think both issues are interesting features and they already implemented:

* JEXL supports the creation of JSON object from measures: https://github.com/telefonicaid/iotagent-node-lib/blob/2e4c113ee02e124ad68f3b58ed950bad999c1aad/test/unit/expressions/jexlExpression-test.js#L253
 If don't, it should. We don't need a new macro/expression language for this, just improve how JEXL behaves (iff you encounter any problem, the use case was one of the drivers for the new JEXL expression plugin).

* "delete" or ignore measures is also implemented by https://github.com/telefonicaid/iotagent-node-lib/pull/873 I think it cover the required use case (if don't it should be fixed/extended).

**Comment by jason-fox** [comment]
So it looks like there is already a mechanism to do the tasks described in the simple plugins proposed, but the current documentation is a bit thin and could do with expanding. To be honest, I didn't look into the unit tests to see if this was implemented, I just tried to read the docs - and in the end it is easier just to  raise an issue and ask to find out.

On reflection for the **string-to-object** thingy it _may_ make sense to be added as a PR option to the legacy expression stuff and I'd definitely like to see more examples for expressions and `explicitAttrs` - may be in a tutorial?

Anyway this was just a  start as I  have a very complex mapping problem, so I'm going to have to write a custom plugin anyway. so I'll still need to do the work myself. For example this is `LineString` attribute in *XML* :smile: 

```xml
<PFD A="PFD3" C="hill" D="32000" G="CTP3">
    <PLN A="1" B="Field Boundary" C="32000" D="1">
      <LSG A="1" B="Line1" C="20" D="280000" E="1">
        <PNT A="2" B="start" C="40.84982" D="-96.596045" E="150234" F="1" />
        <PNT A="2" B="mid" C="40.84982" D="-96.592655" E="148987" F="1" />
        <PNT A="2" B="end" C="40.846573" D="-96.592526" E="148284" F="1" />
      </LSG>
    </PLN>
    <GGP A="GGP1">
      <GPN A="GPN1" C="1" D="1">
        <LSG A="5" B="Guidance1" C="6000" D="1500000" E="1">
          <PNT A="6" B="start" C="40.84934" D="-96.593445" E="148987" F="1" H="0.05" />
          <PNT A="7" B="end" C="40.84683" D="-96.592225" E="150234" F="1" H="0.04" />
        </LSG>
      </GPN>
      <GPN A="GPN2" C="1" D="1">
        <LSG A="5" E="1" B="Guidance2" D="900000" C="18000" F="LSG2">
          <PNT A="2" B="start" C="40.84934" D="-96.122322" E="123506" F="1" I="2.0" />
          <PNT A="2" B="end" C="40.84911" D="-96.591122" E="122544" F="1" I="2.0" />
        </LSG>
      </GPN>
    </GGP>
  </PFD>
  <TSK A="TSK1" F="WKR1" G="1">
    <GAN A="GGP1">
      <ASP A="2003-08-20T08:00:00" B="2003-08-20T17:00:00" D="1" />
    </GAN>
  </TSK>
```

and the *XML* _"measure"_ above should end up as three separate entities - something like this:

```json
{
    "id": "urn:ngsi-ld:AgriParcel:PFD3",
    "type": "AgriParcel",
    "Name": "Hill",
...etc    
}
```

```json
{
    "id" : "urn:ngsi-ld:AgriPlan:GGP1",
    "type": "AgriPlan",
    "name": "Guidance1",
	"Guidance": [
			[[40.84982,-96.596045], [40.84982,-96.592655],[40.846573,-96.592526]],
			[[40.84982,-96.596045], [40.84982,-96.592655],[40.846573,-96.592526]]
	]
}
```

```json
{
    "id": "urn:ngsi-ld:Task:TSK1",
    "type": "Task",
    "refWorker": "urn:ngsi-ld:Person:WKR1",
    "refGuidancePlan": "urn:ngsi-ld:AgriPlan:GGP1",
    "partField" : "urn:ngsi-ld:AgriParcel:PFD3"
}
```
Try mapping that using the expression plugin! 😄

**Comment by mrutid** [comment]
The JEXL expression plugin is already documented, I just pointed out the test to make clear the case you were talking about is currently covered by the language. Now i see the initial use case has changed a little bit... For that kind of NEW measure you don't need a plugin, you'll need a new IoT Agent since it exceed the current JSON or UL scopes. Sure an XML agent with XPath expressions would be useful for that kind of measures.

I'm almost sure there is also documentation for the explicit attributes, we are not used to merge without documentation. Like before I just point out the PR/Issue to be sure it was implemented and because it knew were to find it quick.

Said that I agree with you that improve documentation, make tutorials and walk-throughs its one of the most important tasks to do in order to improve the agents or in general any GE. I you find something missing or had any idea for improvement please make an issue.

**Comment by jason-fox** [comment]
> ... For that kind of NEW measure you don't need a plugin, you'll need a new IoT Agent since it exceed the current JSON or UL scopes. Sure an XML agent with XPath expressions would be useful for that kind of measures.
>

Sure - work in progress : https://github.com/jason-fox/iotagent-ADAPT :smile:

**Comment by jason-fox** [comment]
Based on https://github.com/telefonicaid/iotagent-node-lib/issues/879#issuecomment-654814424

* JEXL supports the creation of JSON object from measures
* "delete" or ignore measures is also implemented by #873

Therefore developing an equivalent will be unnecessary

**Comment by chicco785** [comment]
> So it looks like there is already a mechanism to do the tasks described in the simple plugins proposed, but the current documentation is a bit thin and could do with expanding. To be honest, I didn't look into the unit tests to see if this was implemented, I just tried to read the docs - and in the end it is easier just to raise an issue and ask to find out.

@jason-fox some examples were proposed and some tests included. Anyhow, given that the jexl plugin is just a wrapper to support JEXL i think the documentation included is fair enough. More details of things that can be done can be found in the jexl documentation.
